File: app/src/main/java/dev/ragnarok/fenrir/domain/impl/StickersInteractor.java
package dev.ragnarok.fenrir.domain.impl;

import static dev.ragnarok.fenrir.domain.mappers.MapUtil.mapAll;
import static dev.ragnarok.fenrir.util.Utils.listEmptyIfNull;

import android.content.Context;

import androidx.annotation.NonNull;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.model.VKApiSticker;
import dev.ragnarok.fenrir.api.model.VKApiStickerSet;
import dev.ragnarok.fenrir.api.model.VkApiStickersKeywords;
import dev.ragnarok.fenrir.db.interfaces.IStickersStorage;
import dev.ragnarok.fenrir.db.model.entity.StickerSetEntity;
import dev.ragnarok.fenrir.db.model.entity.StickersKeywordsEntity;
import dev.ragnarok.fenrir.domain.IStickersInteractor;
import dev.ragnarok.fenrir.domain.mappers.Dto2Entity;
import dev.ragnarok.fenrir.domain.mappers.Entity2Model;
import dev.ragnarok.fenrir.model.Sticker;
import dev.ragnarok.fenrir.model.StickerSet;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.AppPerms;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;


public class StickersInteractor implements IStickersInteractor {

    private final INetworker networker;
    private final IStickersStorage storage;

    public StickersInteractor(INetworker networker, IStickersStorage storage) {
        this.networker = networker;
        this.storage = storage;
    }

    @Override
    public Completable getAndStore(int accountId) {
        Completable stickerSet = networker.vkDefault(accountId)
                .store()
                .getStickers()
                .flatMapCompletable(items -> {
                    List<VKApiStickerSet.Product> list = listEmptyIfNull(items.sticker_pack.items);

                    if (Settings.get().ui().isStickers_by_new()) {
                        Collections.reverse(list);
                    }

                    StickerSetEntity temp = new StickerSetEntity(-1).setTitle("recent")
                            .setStickers(mapAll(listEmptyIfNull(listEmptyIfNull(items.recent.items)), Dto2Entity::mapSticker)).setActive(true).setPurchased(true);
                    List<StickerSetEntity> ret = mapAll(list, Dto2Entity::mapStikerSet);
                    ret.add(temp);
                    return storage.store(accountId, ret);
                });
        if (Settings.get().other().isHint_stickers()) {
            return stickerSet.andThen(networker.vkDefault(accountId)
                    .store()
                    .getStickerKeywords()
                    .flatMapCompletable(hint -> getStickersKeywordsAndStore(accountId, hint)));
        }
        return stickerSet;
    }

    private List<StickersKeywordsEntity> generateKeywords(@NonNull List<List<VKApiSticker>> s, @NonNull List<List<String>> w) {
        List<StickersKeywordsEntity> ret = new ArrayList<>(w.size());
        for (int i = 0; i < w.size(); i++) {
            if (Utils.isEmpty(w.get(i))) {
                continue;
            }
            ret.add(new StickersKeywordsEntity(w.get(i), mapAll(listEmptyIfNull(s.get(i)), Dto2Entity::mapSticker)));
        }
        return ret;
    }

    private Completable getStickersKeywordsAndStore(int accountId, VkApiStickersKeywords items) {
        List<List<VKApiSticker>> s = listEmptyIfNull(items.words_stickers);
        List<List<String>> w = listEmptyIfNull(items.keywords);
        List<StickersKeywordsEntity> temp = new ArrayList<>();
        if (Utils.isEmpty(w) || Utils.isEmpty(s) || w.size() != s.size()) {
            return storage.storeKeyWords(accountId, temp);
        }

        temp.addAll(generateKeywords(s, w));
        return storage.storeKeyWords(accountId, temp);
    }

    @Override
    public Single<List<StickerSet>> getStickers(int accountId) {
        return storage.getPurchasedAndActive(accountId)
                .map(entities -> mapAll(entities, Entity2Model::map));
    }

    @Override
    public Single<List<Sticker>> getKeywordsStickers(int accountId, String s) {
        return storage.getKeywordsStickers(accountId, s)
                .map(entities -> mapAll(entities, Entity2Model::buildStickerFromDbo));
    }

    @Override
    public Completable PlaceToStickerCache(Context context) {
        if (!AppPerms.hasReadStoragePermissionSimple(context))
            return Completable.complete();
        return Completable.create(t -> {
            File temp = new File(Settings.get().other().getStickerDir());
            if (!temp.exists()) {
                t.onComplete();
                return;
            }
            File[] file_list = temp.listFiles();
            if (file_list == null || file_list.length <= 0) {
                t.onComplete();
                return;
            }
            Arrays.sort(file_list, (o1, o2) -> Long.compare(o2.lastModified(), o1.lastModified()));
            Utils.getCachedMyStickers().clear();
            for (File u : file_list) {
                if (u.isFile() && (u.getName().contains(".png") || u.getName().contains(".webp"))) {
                    Utils.getCachedMyStickers().add(new Sticker.LocalSticker(u.getAbsolutePath(), false));
                } else if (u.isFile() && u.getName().contains(".json")) {
                    Utils.getCachedMyStickers().add(new Sticker.LocalSticker(u.getAbsolutePath(), true));
                }
            }
        });
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/domain/impl/StickersInteractor.kt
package dev.ragnarok.fenrir.domain.impl

import android.content.Context
import dev.ragnarok.fenrir.api.interfaces.INetworker
import dev.ragnarok.fenrir.api.model.VKApiSticker
import dev.ragnarok.fenrir.api.model.VKApiStickerSet.Product
import dev.ragnarok.fenrir.api.model.VkApiStickerSetsData
import dev.ragnarok.fenrir.api.model.VkApiStickersKeywords
import dev.ragnarok.fenrir.db.interfaces.IStickersStorage
import dev.ragnarok.fenrir.db.model.entity.StickerSetEntity
import dev.ragnarok.fenrir.db.model.entity.StickersKeywordsEntity
import dev.ragnarok.fenrir.domain.IStickersInteractor
import dev.ragnarok.fenrir.domain.mappers.Dto2Entity.mapSticker
import dev.ragnarok.fenrir.domain.mappers.Dto2Entity.mapStikerSet
import dev.ragnarok.fenrir.domain.mappers.Entity2Model.buildStickerFromDbo
import dev.ragnarok.fenrir.domain.mappers.Entity2Model.map
import dev.ragnarok.fenrir.domain.mappers.MapUtil.mapAll
import dev.ragnarok.fenrir.domain.mappers.MapUtil.mapAllMutable
import dev.ragnarok.fenrir.model.Sticker
import dev.ragnarok.fenrir.model.Sticker.LocalSticker
import dev.ragnarok.fenrir.model.StickerSet
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.AppPerms.hasReadStoragePermissionSimple
import dev.ragnarok.fenrir.util.Utils.getCachedMyStickers
import dev.ragnarok.fenrir.util.Utils.listEmptyIfNull
import dev.ragnarok.fenrir.util.Utils.listEmptyIfNullMutable
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single
import java.io.File
import java.util.*

class StickersInteractor(private val networker: INetworker, private val storage: IStickersStorage) :
    IStickersInteractor {
    override fun getAndStore(accountId: Int): Completable {
        val stickerSet = networker.vkDefault(accountId)
            .store()
            .stickers
            .flatMapCompletable { items: VkApiStickerSetsData? ->
                val list: MutableList<Product> = listEmptyIfNullMutable(items?.sticker_pack?.items)
                if (Settings.get().ui().isStickers_by_new) {
                    list.reverse()
                }
                val temp = StickerSetEntity(-1).setTitle("recent")
                    .setStickers(mapAll(listEmptyIfNull(listEmptyIfNull(items?.recent?.items))) {
                        mapSticker(
                            it
                        )
                    }).setActive(true).setPurchased(true)
                val ret =
                    mapAllMutable(list) { mapStikerSet(it) }
                ret.add(temp)
                storage.store(accountId, ret)
            }
        return if (Settings.get().other().isHint_stickers) {
            stickerSet.andThen(
                networker.vkDefault(accountId)
                    .store()
                    .stickerKeywords
                    .flatMapCompletable { hint: VkApiStickersKeywords ->
                        getStickersKeywordsAndStore(
                            accountId,
                            hint
                        )
                    })
        } else stickerSet
    }

    private fun generateKeywords(
        s: List<List<VKApiSticker>?>,
        w: List<List<String?>?>
    ): List<StickersKeywordsEntity> {
        val ret: MutableList<StickersKeywordsEntity> = ArrayList(w.size)
        for (i in w.indices) {
            if (w[i].isNullOrEmpty()) {
                continue
            }
            ret.add(
                StickersKeywordsEntity(
                    w[i], mapAll(
                        listEmptyIfNull(
                            s[i]
                        )
                    ) { mapSticker(it) })
            )
        }
        return ret
    }

    private fun getStickersKeywordsAndStore(
        accountId: Int,
        items: VkApiStickersKeywords
    ): Completable {
        val s: List<List<VKApiSticker>> = listEmptyIfNull(items.words_stickers)
        val w: List<List<String>> = listEmptyIfNull(items.keywords)
        val temp: MutableList<StickersKeywordsEntity> = ArrayList()
        if (w.isNullOrEmpty() || s.isNullOrEmpty() || w.size != s.size) {
            return storage.storeKeyWords(accountId, temp)
        }
        temp.addAll(generateKeywords(s, w))
        return storage.storeKeyWords(accountId, temp)
    }

    override fun getStickers(accountId: Int): Single<List<StickerSet>> {
        return storage.getPurchasedAndActive(accountId)
            .map { entities ->
                mapAll(entities) {
                    map(
                        it
                    )
                }
            }
    }

    override fun getKeywordsStickers(accountId: Int, s: String?): Single<List<Sticker>> {
        return storage.getKeywordsStickers(accountId, s)
            .map { entities ->
                mapAll(entities) {
                    buildStickerFromDbo(
                        it
                    )
                }
            }
    }

    override fun PlaceToStickerCache(context: Context): Completable {
        return if (!hasReadStoragePermissionSimple(context)) Completable.complete() else Completable.create { t ->
            val temp = File(Settings.get().other().stickerDir)
            if (!temp.exists()) {
                t.onComplete()
                return@create
            }
            val file_list = temp.listFiles()
            if (file_list == null || file_list.isEmpty()) {
                t.onComplete()
                return@create
            }
            Arrays.sort(file_list) { o1: File, o2: File ->
                o2.lastModified().compareTo(o1.lastModified())
            }
            getCachedMyStickers().clear()
            for (u in file_list) {
                if (u.isFile && (u.name.contains(".png") || u.name.contains(".webp"))) {
                    getCachedMyStickers().add(LocalSticker(u.absolutePath, false))
                } else if (u.isFile && u.name.contains(".json")) {
                    getCachedMyStickers().add(LocalSticker(u.absolutePath, true))
                }
            }
        }
    }
}